File: src/module.ml
let _ = Callback.register "array_of_list" Array.of_list
let _ = Callback.register "array_length" Array.length

type t

module Feature = struct
  type t = int

  external mvp : unit -> t = "caml_binaryen_feature_mvp"

  let mvp = mvp ()

  external atomics : unit -> t = "caml_binaryen_feature_atomics"

  let atomics = atomics ()

  external mutable_globals : unit -> t = "caml_binaryen_feature_mutable_globals"

  let mutable_globals = mutable_globals ()

  external nontrapping_fp_to_int : unit -> t
    = "caml_binaryen_feature_nontrapping_fp_to_int"

  let nontrapping_fp_to_int = nontrapping_fp_to_int ()

  external simd128 : unit -> t = "caml_binaryen_feature_simd128"

  let simd128 = simd128 ()

  external bulk_memory : unit -> t = "caml_binaryen_feature_bulk_memory"

  let bulk_memory = bulk_memory ()

  external sign_ext : unit -> t = "caml_binaryen_feature_sign_ext"

  let sign_ext = sign_ext ()

  external exception_handling : unit -> t
    = "caml_binaryen_feature_exception_handling"

  let exception_handling = exception_handling ()

  external tail_call : unit -> t = "caml_binaryen_feature_tail_call"

  let tail_call = tail_call ()

  external reference_types : unit -> t = "caml_binaryen_feature_reference_types"

  let reference_types = reference_types ()

  external multivalue : unit -> t = "caml_binaryen_feature_multivalue"

  let multivalue = multivalue ()

  external gc : unit -> t = "caml_binaryen_feature_gc"

  let gc = gc ()

  external memory64 : unit -> t = "caml_binaryen_feature_memory64"

  let memory64 = memory64 ()

  external relaxed_simd : unit -> t = "caml_binaryen_feature_relaxed_simd"

  let relaxed_simd = relaxed_simd ()

  external extended_const : unit -> t = "caml_binaryen_feature_extended_const"

  let extended_const = extended_const ()

  external strings : unit -> t = "caml_binaryen_feature_strings"

  let strings = strings ()

  external multi_memory : unit -> t = "caml_binaryen_feature_multi_memory"

  let multi_memory = multi_memory ()

  external stack_switching : unit -> t = "caml_binaryen_feature_stack_switching"

  let stack_switching = stack_switching ()

  external shared_everything : unit -> t
    = "caml_binaryen_feature_shared_everything"

  let shared_everything = shared_everything ()

  external fp16 : unit -> t = "caml_binaryen_feature_fp16"

  let fp16 = fp16 ()

  external bulk_memory_opt : unit -> t = "caml_binaryen_feature_bulk_memory_opt"

  let bulk_memory_opt = bulk_memory_opt ()

  external call_indirect_overlong : unit -> t
    = "caml_binaryen_feature_call_indirect_overlong"

  let call_indirect_overlong = call_indirect_overlong ()

  external all : unit -> t = "caml_binaryen_feature_all"

  let all = all ()
end

external create : unit -> t = "caml_binaryen_module_create"
external dispose : t -> unit = "caml_binaryen_module_dispose"

external add_custom_section : t -> string -> string -> unit
  = "caml_binaryen_add_custom_section"

external parse : string -> t = "caml_binaryen_module_parse"
external print : t -> unit = "caml_binaryen_module_print"
external print_asmjs : t -> unit = "caml_binaryen_module_print_asmjs"
external print_stack_ir : t -> unit = "caml_binaryen_module_print_stack_ir"
external validate : t -> int = "caml_binaryen_module_validate"
external optimize : t -> unit = "caml_binaryen_module_optimize"
external get_features : t -> int = "caml_binaryen_module_get_features"

let get_features wasm_mod =
  let features_int = get_features wasm_mod in
  let rec split_features = function
    | 0 -> []
    | feature when features_int land feature > 0 ->
        feature :: split_features (feature lsr 1)
    | feature -> split_features (feature lsr 1)
  in
  (* Support 32-bit OCaml where integers are 31 bits *)
  (* This supports up to 31 Binaryen features *)
  split_features 0x40000000

external set_features : t -> int -> unit = "caml_binaryen_module_set_features"

let set_features wasm_mod features =
  set_features wasm_mod (List.fold_left ( lor ) 0 features)

external run_passes : t -> Passes.t list -> unit
  = "caml_binaryen_module_run_passes"

external write : t -> string option -> bytes * string option
  = "caml_binaryen_module_write"

external write_text : t -> string = "caml_binaryen_module_write_text"
external write_stack_ir : t -> string = "caml_binaryen_module_write_stack_ir"
external read : bytes -> t = "caml_binaryen_module_read"
external interpret : t -> unit = "caml_binaryen_module_interpret"

external add_debug_info_filename : t -> string -> int
  = "caml_binaryen_module_add_debug_info_filename"

external get_debug_info_filename : t -> int -> string
  = "caml_binaryen_module_get_debug_info_filename"

external update_maps : t -> unit = "caml_binaryen_module_update_maps"

File: src/module.js
//Provides: caml_binaryen_module_create
//Requires: Binaryen
function caml_binaryen_module_create() {
  return new Binaryen.Module();
}

//Provides: caml_binaryen_module_dispose
function caml_binaryen_module_dispose(wasm_mod) {
  return wasm_mod.dispose();
}

//Provides: caml_binaryen_module_parse
//Requires: Binaryen
//Requires: caml_jsstring_of_string
function caml_binaryen_module_parse(text) {
  return Binaryen.parseText(caml_jsstring_of_string(text));
}

//Provides: caml_binaryen_module_print
//Requires: caml_sys_fds
//Requires: caml_string_of_jsstring
//Requires: caml_ml_output, caml_ml_string_length
function caml_binaryen_module_print(wasm_mod) {
  var text = wasm_mod.emitText();
  var chanid = caml_sys_fds[1].chanid; // stdout
  var s = caml_string_of_jsstring(text);
  caml_ml_output(chanid, s, 0, caml_ml_string_length(s));
}

//Provides: caml_binaryen_module_print_asmjs
//Requires: caml_sys_fds
//Requires: caml_string_of_jsstring
//Requires: caml_ml_output, caml_ml_string_length
function caml_binaryen_module_print_asmjs(wasm_mod) {
  var asm = wasm_mod.emitAsmjs();
  var chanid = caml_sys_fds[1].chanid; // stdout
  var s = caml_string_of_jsstring(asm);
  caml_ml_output(chanid, s, 0, caml_ml_string_length(s));
}

//Provides: caml_binaryen_module_print_stack_ir
//Requires: caml_sys_fds
//Requires: caml_string_of_jsstring
//Requires: caml_ml_output, caml_ml_string_length
function caml_binaryen_module_print_stack_ir(wasm_mod) {
  var stackir = wasm_mod.emitStackIR();
  var chanid = caml_sys_fds[1].chanid; // stdout
  var s = caml_string_of_jsstring(stackir);
  caml_ml_output(chanid, s, 0, caml_ml_string_length(s));
}

//Provides: caml_binaryen_module_validate
function caml_binaryen_module_validate(wasm_mod) {
  return wasm_mod.validate();
}

//Provides: caml_binaryen_module_optimize
function caml_binaryen_module_optimize(wasm_mod) {
  return wasm_mod.optimize();
}

//Provides: caml_binaryen_module_run_passes
//Requires: caml_jsstring_of_string, caml_list_to_js_array
function caml_binaryen_module_run_passes(wasm_mod, passes) {
  return wasm_mod.runPasses(
    caml_list_to_js_array(passes).map(caml_jsstring_of_string)
  );
}

//Provides: caml_binaryen_module_auto_drop
function caml_binaryen_module_auto_drop(wasm_mod) {
  return wasm_mod.autoDrop();
}

//Provides: caml_binaryen_module_write
//Requires: to_option
//Requires: caml_jsstring_of_string, caml_string_of_jsstring, caml_bytes_of_array
function caml_binaryen_module_write(wasm_mod, sourceMapUrl) {
  if (sourceMapUrl) {
    var url = caml_jsstring_of_string(sourceMapUrl[1]);
    var obj = wasm_mod.emitBinary(url);
    var sourceMap =
      obj.sourceMap != null ? caml_string_of_jsstring(obj.sourceMap) : null;
    return [0, caml_bytes_of_array(obj.binary), to_option(sourceMap)];
  } else {
    var binary = wasm_mod.emitBinary();
    return [0, caml_bytes_of_array(binary), 0];
  }
}

//Provides: caml_binaryen_module_write_text
//Requires: caml_string_of_jsstring
function caml_binaryen_module_write_text(wasm_mod) {
  var text = wasm_mod.emitText();
  return caml_string_of_jsstring(text);
}

//Provides: caml_binaryen_module_write_stack_ir
//Requires: caml_string_of_jsstring
function caml_binaryen_module_write_stack_ir(wasm_mod) {
  var text = wasm_mod.emitStackIR();
  return caml_string_of_jsstring(text);
}

//Provides: caml_binaryen_module_read
//Requires: Binaryen
//Requires: caml_uint8_array_of_bytes
function caml_binaryen_module_read(bytes) {
  var data = caml_uint8_array_of_bytes(bytes);
  return Binaryen.readBinary(data);
}

//Provides: caml_binaryen_module_interpret
function caml_binaryen_module_interpret(wasm_mod) {
  return wasm_mod.interpret();
}

//Provides: caml_binaryen_module_add_debug_info_filename
//Requires: caml_jsstring_of_string
function caml_binaryen_module_add_debug_info_filename(wasm_mod, filename) {
  return wasm_mod.addDebugInfoFileName(caml_jsstring_of_string(filename));
}

//Provides: caml_binaryen_module_get_debug_info_filename
function caml_binaryen_module_get_debug_info_filename(wasm_mod, index) {
  return wasm_mod.getDebugInfoFileName(index);
}

//Provides: caml_binaryen_add_custom_section
//Requires: caml_jsstring_of_string, caml_array_of_string
function caml_binaryen_add_custom_section(wasm_mod, name, contents) {
  return wasm_mod.addCustomSection(
    caml_jsstring_of_string(name),
    caml_array_of_string(contents)
  );
}

//Provides: caml_binaryen_module_update_maps
function caml_binaryen_module_update_maps(wasm_mod) {
  // The JS version doesn't expose `update_maps`
  return;
}
